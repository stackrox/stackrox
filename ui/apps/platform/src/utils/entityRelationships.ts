// Entity Relationships

// note: the relationships are directional!
// changing direction may change relationship type between entities!!

import { RelationshipType } from 'constants/relationshipTypes';
import { IsFeatureFlagEnabled } from 'hooks/useFeatureFlags';

/*
// For historical interest: never used.
const complianceEntityTypes = ['CONTROL', 'NODE', 'CLUSTER', 'NAMESPACE', 'DEPLOYMENT'];
*/

const configurationManagementEntityTypes = [
    'CONTROL',
    'NODE',
    'IMAGE',
    'ROLE',
    'SECRET',
    'SUBJECT',
    'SERVICE_ACCOUNT',
    'POLICY',
    'CLUSTER',
    'NAMESPACE',
    'DEPLOYMENT',
] as const; // necessary for type below

export type ConfigurationManagementEntityType = (typeof configurationManagementEntityTypes)[number];

export function getConfigurationManagementEntityTypes(
    isFeatureFlagEnabled?: IsFeatureFlagEnabled
): ConfigurationManagementEntityType[] {
    if (isFeatureFlagEnabled) {
        // Arrays include all possible entity types for use case.
        return configurationManagementEntityTypes.filter((/* entityType */) => {
            /*
            // Pattern to filter out an entity type if a feature flag is not enabled.
            if (entityType === 'WHICHEVER' && !isFeatureFlagEnabled(ROX_WHATEVER)) {
                return false;
            } 
            */
            return true;
        });
    }

    return [...configurationManagementEntityTypes];
}

const vulnerabilityManagementEntityTypes = [
    'IMAGE_CVE',
    'NODE_CVE',
    'CLUSTER_CVE',
    'CLUSTER',
    'NAMESPACE',
    'DEPLOYMENT',
    'IMAGE',
    'IMAGE_COMPONENT',
    'NODE_COMPONENT',
    'NODE',
] as const; // necessary for type below

export type VulnerabilityManagementEntityType = (typeof vulnerabilityManagementEntityTypes)[number];

export function getVulnerabilityManagementEntityTypes(
    isFeatureFlagEnabled?: IsFeatureFlagEnabled
): VulnerabilityManagementEntityType[] {
    if (isFeatureFlagEnabled) {
        // Arrays include all possible entity types for use case.
        return vulnerabilityManagementEntityTypes.filter((/* entityType */) => {
            /*
            // Pattern to filter out an entity type if a feature flag is not enabled.
            if (entityType === 'WHICHEVER' && !isFeatureFlagEnabled(ROX_WHATEVER)) {
                return false;
            } 
            */
            return true;
        });
    }

    return [...vulnerabilityManagementEntityTypes];
}

export type EntityGroup =
    | 'OVERVIEW'
    | 'VIOLATIONS_AND_FINDINGS'
    | 'APPLICATION_RESOURCES'
    | 'RBAC_CONFIG'
    | 'SECURITY';

export const entityGroups: Record<EntityGroup, string> = {
    OVERVIEW: 'Overview',
    VIOLATIONS_AND_FINDINGS: 'Violations & Findings',
    APPLICATION_RESOURCES: 'Application & Infrastructure',
    RBAC_CONFIG: 'Role-Based Access Control',
    SECURITY: 'Security Findings',
};

export type EntityType = ConfigurationManagementEntityType | VulnerabilityManagementEntityType;

export const entityGroupMap: Record<EntityType, EntityGroup> = {
    ROLE: 'RBAC_CONFIG',
    SUBJECT: 'RBAC_CONFIG',
    SERVICE_ACCOUNT: 'RBAC_CONFIG',

    DEPLOYMENT: 'APPLICATION_RESOURCES',
    SECRET: 'APPLICATION_RESOURCES',
    NODE: 'APPLICATION_RESOURCES',
    CLUSTER: 'APPLICATION_RESOURCES',
    NAMESPACE: 'APPLICATION_RESOURCES',
    IMAGE: 'APPLICATION_RESOURCES',
    NODE_COMPONENT: 'APPLICATION_RESOURCES',
    IMAGE_COMPONENT: 'APPLICATION_RESOURCES',

    POLICY: 'SECURITY',
    CONTROL: 'SECURITY',
    NODE_CVE: 'SECURITY',
    IMAGE_CVE: 'SECURITY',
    CLUSTER_CVE: 'SECURITY',
};

type EntityRelationshipData = {
    children: EntityType[];
    parents: EntityType[];
    matches: EntityType[];
    extendedMatches?: EntityType[];
};

// If you change the data, then you will need to update a snapshot.
const entityRelationshipMap: Record<EntityType, EntityRelationshipData> = {
    CLUSTER: {
        children: ['NODE', 'NAMESPACE', 'ROLE'],
        parents: [],
        matches: ['CONTROL', 'CLUSTER_CVE'],
        // extendedMatches: [entityTypes.POLICY]
    },
    NODE: {
        children: ['NODE_COMPONENT'],
        parents: ['CLUSTER'],
        matches: ['CONTROL'],
    },
    NAMESPACE: {
        children: ['DEPLOYMENT', 'SERVICE_ACCOUNT', 'SECRET'],
        parents: ['CLUSTER'],
        matches: [],
        // extendedMatches: [entityTypes.POLICY]
    },
    DEPLOYMENT: {
        children: ['IMAGE'],
        parents: ['NAMESPACE', 'CLUSTER'],
        matches: ['SERVICE_ACCOUNT', 'POLICY', 'CONTROL', 'SECRET'],
    },
    IMAGE: {
        children: ['IMAGE_COMPONENT'],
        parents: [],
        matches: ['DEPLOYMENT'],
    },
    NODE_COMPONENT: {
        children: [],
        parents: [],
        matches: ['NODE_CVE', 'NODE'],
        extendedMatches: [],
    },
    IMAGE_COMPONENT: {
        children: [],
        parents: [],
        matches: ['IMAGE', 'IMAGE_CVE'],
        extendedMatches: ['DEPLOYMENT'],
    },
    IMAGE_CVE: {
        children: [],
        parents: [],
        matches: ['IMAGE_COMPONENT'],
        extendedMatches: ['IMAGE', 'DEPLOYMENT'],
    },
    NODE_CVE: {
        children: [],
        parents: [],
        matches: ['NODE_COMPONENT'],
        extendedMatches: ['NODE'],
    },
    CLUSTER_CVE: {
        children: [],
        parents: [],
        matches: [],
        extendedMatches: ['CLUSTER'],
    },
    CONTROL: {
        children: [],
        parents: [],
        matches: ['NODE', 'DEPLOYMENT', 'CLUSTER'],
    },
    POLICY: {
        children: [],
        parents: [],
        matches: ['DEPLOYMENT'],
    },
    SECRET: {
        children: [],
        parents: ['NAMESPACE'],
        matches: ['DEPLOYMENT'],
    },
    SUBJECT: {
        children: [],
        parents: [],
        matches: ['ROLE'],
    },
    SERVICE_ACCOUNT: {
        children: [],
        parents: ['NAMESPACE'],
        matches: ['DEPLOYMENT', 'ROLE'],
    },
    ROLE: {
        children: [],
        parents: ['CLUSTER'],
        matches: ['SERVICE_ACCOUNT', 'SUBJECT'],
    },
};

// helper functions
const getChildren = (entityType: EntityType): EntityType[] =>
    entityRelationshipMap[entityType].children;
const getParents = (entityType: EntityType): EntityType[] =>
    entityRelationshipMap[entityType].parents;
const getPureMatches = (entityType: EntityType): EntityType[] =>
    entityRelationshipMap[entityType].matches;
const getExtendedMatches = (entityType: EntityType): EntityType[] =>
    entityRelationshipMap[entityType].extendedMatches || [];
const getMatches = (entityType: EntityType): EntityType[] => [
    ...getPureMatches(entityType),
    ...getExtendedMatches(entityType),
];

// function to recursively get inclusive 'contains' relationships (inferred)
// this includes all generations of children AND inferred (matches of children down the chain) relationships
// e.g. namespace inclusively contains policy since ns contains deployment and deployment matches policy
const getContains = (entityType: EntityType): EntityType[] => {
    const relationships: EntityType[] = [];

    function pushChild(child) {
        // Do not include itself. Prevent duplicates.
        if (child !== entityType && !relationships.includes(child)) {
            relationships.push(child);
        }
    }

    getChildren(entityType).forEach((child) => {
        pushChild(child);
        getPureMatches(child).forEach((childMatches) => {
            pushChild(childMatches);
        });
        getContains(child).forEach((childContains) => {
            pushChild(childContains);
        });
    });

    return relationships;
};

const isChild = (parent: EntityType, child: EntityType): boolean =>
    getChildren(parent).includes(child);
const isParent = (parent: EntityType, child: EntityType): boolean =>
    getParents(child).includes(parent);
const isMatch = (entityType1: EntityType, entityType2: EntityType): boolean =>
    getMatches(entityType1).includes(entityType2);
const isPureMatch = (entityType1: EntityType, entityType2: EntityType): boolean =>
    getPureMatches(entityType1).includes(entityType2);
const isExtendedMatch = (entityType1: EntityType, entityType2: EntityType): boolean =>
    getExtendedMatches(entityType1).includes(entityType2);
const isContained = (entityType1: EntityType, entityType2: EntityType): boolean =>
    getContains(entityType1).includes(entityType2);
const isContainedInferred = (entityType1: EntityType, entityType2: EntityType): boolean =>
    entityType1 !== entityType2 &&
    isContained(entityType1, entityType2) &&
    !isChild(entityType1, entityType2);

function getEntityTypesByRelationship(
    entityType: EntityType,
    relationship: RelationshipType
): EntityType[] {
    switch (relationship) {
        case 'CONTAINS': {
            const entities = getContains(entityType);
            // this is to remove NODE links from IMAGE, DEPLOYMENT, NAMESPACE and vice versa
            // need to revisit the mapping later.
            if (entityType === 'NODE') {
                return entities.filter((entity) => entity !== 'IMAGE');
            }
            if (
                entityType === 'IMAGE' ||
                entityType === 'DEPLOYMENT' ||
                entityType === 'NAMESPACE'
            ) {
                return entities.filter((entity) => entity !== 'NODE');
            }
            return entities;
        }
        case 'MATCHES':
            return getMatches(entityType);
        case 'PARENTS': // not used in UI
            return getParents(entityType);
        case 'CHILDREN': // not used in UI
            return getChildren(entityType);
        default:
            return [];
    }
}

// Configuration Management never used this method to compute its EntityTypesByRelationship.

export function getVulnerabilityManagementEntityTypesByRelationship(
    entityType: EntityType,
    relationship: RelationshipType,
    isFeatureFlagEnabled?: IsFeatureFlagEnabled
) {
    const entityTypes = getVulnerabilityManagementEntityTypes(isFeatureFlagEnabled);

    // Do not include itself. Filter out any Configuration Management entity types.
    return getEntityTypesByRelationship(entityType, relationship).filter(
        (entityTypeByRelationship) =>
            entityTypeByRelationship !== entityType &&
            entityTypes.includes(entityTypeByRelationship as VulnerabilityManagementEntityType)
    ) as VulnerabilityManagementEntityType[];
}

export default {
    getChildren,
    getParents,
    getMatches,
    getContains,
    isChild,
    isParent,
    isMatch,
    isPureMatch,
    isExtendedMatch,
    isContained,
    isContainedInferred,
};
