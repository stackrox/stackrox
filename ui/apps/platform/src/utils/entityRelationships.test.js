import intersection from 'lodash/intersection';

import entityRelationships, {
    entityGroupMap,
    getVulnerabilityManagementEntityTypesByRelationship as getEntityTypesByRelationship,
} from './entityRelationships';

describe('entityRelationshipMap', () => {
    // Assume that entityGroupMap has all entity types which have relationships.
    const entityTypesRelationships = Object.keys(entityGroupMap).sort();

    /**
     * Emulate get*s function that is not exported
     * using is* callback function that is exported.
     * The returned array is sorted, because entityTypes is sorted.
     */
    const enumerateTypes = (entityType1, callback) =>
        entityTypesRelationships.filter((entityType2) => callback(entityType1, entityType2));

    const {
        getChildren,
        getContains,
        getMatches,
        getParents,
        isContainedInferred,
        isExtendedMatch,
        isPureMatch,
    } = entityRelationships;

    const receivedRelationships = {};
    entityTypesRelationships.forEach((entityType) => {
        const children = [...getChildren(entityType)].sort(); // copy before sort
        const parents = [...getParents(entityType)].sort();
        const matches = enumerateTypes(entityType, isPureMatch);
        const extendedMatches = enumerateTypes(entityType, isExtendedMatch);
        const expectedMatches = [...matches, ...extendedMatches].sort();
        const receivedMatches = [...getMatches(entityType)].sort();
        const receivedContains = [...getContains(entityType)].sort();
        const receivedInferred = enumerateTypes(entityType, isContainedInferred);

        // Numbers in keys for order to compare with entityRelationshipMap.
        // Spaces in keys for alignment to compare related relationships.
        receivedRelationships[entityType] = {
            '1-children': children.join(' '),
            '2-parents': parents.join(' '),
            '3-matches        ': matches.join(' '),
            '4-extendedMatches': extendedMatches.join(' '),
            '5-getMatches     ': receivedMatches.join(' '),
            '6-getContains        ': receivedContains.join(' '),
            '7-isContainedInferred': receivedInferred.join(' '),
        };

        // Template literal for jest/valid-title which forbids a variable.
        describe(`${entityType}`, () => {
            it('has relationships that are not reflexive', () => {
                expect(children).not.toContain(entityType);
                expect(parents).not.toContain(entityType);
                expect(expectedMatches).not.toContain(entityType);
                expect(receivedContains).not.toContain(entityType);
                expect(receivedInferred).not.toContain(entityType);
            });

            it(`has properties that are disjoint`, () => {
                expect(intersection(children, parents, matches, extendedMatches)).toEqual([]);
            });

            it(`has getMatches method that is union of matches and extendedMatches`, () => {
                expect(receivedMatches).toEqual(expectedMatches);
            });

            it(`has isContained method is union of children and isContainedInferred`, () => {
                const expectedContains = [...children, ...receivedInferred].sort();
                expect(receivedContains).toEqual(expectedContains);
            });
        });
    });

    // Snapshot shows the effect of changes to data or logic.
    it('has properties and derived properties', () => {
        expect(receivedRelationships).toMatchSnapshot(); // not filtered by use case
    });
});

it('gets Matches relationships for Cluster', () => {
    const matchesForCluster = getEntityTypesByRelationship('CLUSTER', 'MATCHES');
    expect(matchesForCluster).not.toContain('CONTROL');
});

it('gets Matches relationships for Image CVE', () => {
    // should also get extended matches
    const matchesForImageCVE = getEntityTypesByRelationship('IMAGE_CVE', 'MATCHES');
    expect(matchesForImageCVE).toContain('IMAGE');
    expect(matchesForImageCVE).toContain('IMAGE_COMPONENT');
    expect(matchesForImageCVE).toContain('DEPLOYMENT');
    expect(matchesForImageCVE).not.toContain('NAMESPACE');
});

it('gets direct Children relationships for Cluster', () => {
    const childrenOfCluster = getEntityTypesByRelationship('CLUSTER', 'CHILDREN');
    expect(childrenOfCluster).toContain('NODE');
    expect(childrenOfCluster).toContain('NAMESPACE');
    expect(childrenOfCluster).not.toContain('ROLE');
});

it('gets Contains relationships for Cluster', () => {
    const containsForCluster = getEntityTypesByRelationship('CLUSTER', 'CONTAINS');

    // should have direct child
    expect(containsForCluster).not.toContain('ROLE');
    expect(containsForCluster).toContain('NAMESPACE');
    // should have direct child's matches
    expect(containsForCluster).not.toContain('SERVICE_ACCOUNT');
    // should have grandchild
    expect(containsForCluster).toContain('DEPLOYMENT');
    // should have grandchild's matches
    expect(containsForCluster).not.toContain('CONTROL');
});

it('gets Contains relationships for Image', () => {
    const containsForImage = getEntityTypesByRelationship('IMAGE', 'CONTAINS');
    // should have direct child
    expect(containsForImage).toContain('IMAGE_COMPONENT');
    // should not contain itself
    expect(containsForImage).not.toContain('IMAGE');
    // should have grandchild
    expect(containsForImage).toContain('IMAGE_CVE');
    // should not have granchild's extended matches
    expect(containsForImage).not.toContain('DEPLOYMENT');
});
