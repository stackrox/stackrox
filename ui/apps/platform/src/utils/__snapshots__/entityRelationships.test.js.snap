// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entityRelationshipMap has properties and derived properties 1`] = `
Object {
  "CLUSTER": Object {
    "1-children": "NAMESPACE NODE ROLE",
    "2-parents": "",
    "3-matches        ": "CLUSTER_CVE CONTROL",
    "4-extendedMatches": "",
    "5-getMatches     ": "CLUSTER_CVE CONTROL",
    "6-getContains        ": "CONTROL DEPLOYMENT IMAGE IMAGE_COMPONENT IMAGE_CVE NAMESPACE NODE NODE_COMPONENT NODE_CVE POLICY ROLE SECRET SERVICE_ACCOUNT SUBJECT",
    "7-isContainedInferred": "CONTROL DEPLOYMENT IMAGE IMAGE_COMPONENT IMAGE_CVE NODE_COMPONENT NODE_CVE POLICY SECRET SERVICE_ACCOUNT SUBJECT",
  },
  "CLUSTER_CVE": Object {
    "1-children": "",
    "2-parents": "",
    "3-matches        ": "",
    "4-extendedMatches": "CLUSTER",
    "5-getMatches     ": "CLUSTER",
    "6-getContains        ": "",
    "7-isContainedInferred": "",
  },
  "CONTROL": Object {
    "1-children": "",
    "2-parents": "",
    "3-matches        ": "CLUSTER DEPLOYMENT NODE",
    "4-extendedMatches": "",
    "5-getMatches     ": "CLUSTER DEPLOYMENT NODE",
    "6-getContains        ": "",
    "7-isContainedInferred": "",
  },
  "DEPLOYMENT": Object {
    "1-children": "IMAGE",
    "2-parents": "CLUSTER NAMESPACE",
    "3-matches        ": "CONTROL POLICY SECRET SERVICE_ACCOUNT",
    "4-extendedMatches": "",
    "5-getMatches     ": "CONTROL POLICY SECRET SERVICE_ACCOUNT",
    "6-getContains        ": "IMAGE IMAGE_COMPONENT IMAGE_CVE",
    "7-isContainedInferred": "IMAGE_COMPONENT IMAGE_CVE",
  },
  "IMAGE": Object {
    "1-children": "IMAGE_COMPONENT",
    "2-parents": "",
    "3-matches        ": "DEPLOYMENT",
    "4-extendedMatches": "",
    "5-getMatches     ": "DEPLOYMENT",
    "6-getContains        ": "IMAGE_COMPONENT IMAGE_CVE",
    "7-isContainedInferred": "IMAGE_CVE",
  },
  "IMAGE_COMPONENT": Object {
    "1-children": "",
    "2-parents": "",
    "3-matches        ": "IMAGE IMAGE_CVE",
    "4-extendedMatches": "DEPLOYMENT",
    "5-getMatches     ": "DEPLOYMENT IMAGE IMAGE_CVE",
    "6-getContains        ": "",
    "7-isContainedInferred": "",
  },
  "IMAGE_CVE": Object {
    "1-children": "",
    "2-parents": "",
    "3-matches        ": "IMAGE_COMPONENT",
    "4-extendedMatches": "DEPLOYMENT IMAGE",
    "5-getMatches     ": "DEPLOYMENT IMAGE IMAGE_COMPONENT",
    "6-getContains        ": "",
    "7-isContainedInferred": "",
  },
  "NAMESPACE": Object {
    "1-children": "DEPLOYMENT SECRET SERVICE_ACCOUNT",
    "2-parents": "CLUSTER",
    "3-matches        ": "",
    "4-extendedMatches": "",
    "5-getMatches     ": "",
    "6-getContains        ": "CONTROL DEPLOYMENT IMAGE IMAGE_COMPONENT IMAGE_CVE POLICY ROLE SECRET SERVICE_ACCOUNT",
    "7-isContainedInferred": "CONTROL IMAGE IMAGE_COMPONENT IMAGE_CVE POLICY ROLE",
  },
  "NODE": Object {
    "1-children": "NODE_COMPONENT",
    "2-parents": "CLUSTER",
    "3-matches        ": "CONTROL",
    "4-extendedMatches": "",
    "5-getMatches     ": "CONTROL",
    "6-getContains        ": "NODE_COMPONENT NODE_CVE",
    "7-isContainedInferred": "NODE_CVE",
  },
  "NODE_COMPONENT": Object {
    "1-children": "",
    "2-parents": "",
    "3-matches        ": "NODE NODE_CVE",
    "4-extendedMatches": "",
    "5-getMatches     ": "NODE NODE_CVE",
    "6-getContains        ": "",
    "7-isContainedInferred": "",
  },
  "NODE_CVE": Object {
    "1-children": "",
    "2-parents": "",
    "3-matches        ": "NODE_COMPONENT",
    "4-extendedMatches": "NODE",
    "5-getMatches     ": "NODE NODE_COMPONENT",
    "6-getContains        ": "",
    "7-isContainedInferred": "",
  },
  "POLICY": Object {
    "1-children": "",
    "2-parents": "",
    "3-matches        ": "DEPLOYMENT",
    "4-extendedMatches": "",
    "5-getMatches     ": "DEPLOYMENT",
    "6-getContains        ": "",
    "7-isContainedInferred": "",
  },
  "ROLE": Object {
    "1-children": "",
    "2-parents": "CLUSTER",
    "3-matches        ": "SERVICE_ACCOUNT SUBJECT",
    "4-extendedMatches": "",
    "5-getMatches     ": "SERVICE_ACCOUNT SUBJECT",
    "6-getContains        ": "",
    "7-isContainedInferred": "",
  },
  "SECRET": Object {
    "1-children": "",
    "2-parents": "NAMESPACE",
    "3-matches        ": "DEPLOYMENT",
    "4-extendedMatches": "",
    "5-getMatches     ": "DEPLOYMENT",
    "6-getContains        ": "",
    "7-isContainedInferred": "",
  },
  "SERVICE_ACCOUNT": Object {
    "1-children": "",
    "2-parents": "NAMESPACE",
    "3-matches        ": "DEPLOYMENT ROLE",
    "4-extendedMatches": "",
    "5-getMatches     ": "DEPLOYMENT ROLE",
    "6-getContains        ": "",
    "7-isContainedInferred": "",
  },
  "SUBJECT": Object {
    "1-children": "",
    "2-parents": "",
    "3-matches        ": "ROLE",
    "4-extendedMatches": "",
    "5-getMatches     ": "ROLE",
    "6-getContains        ": "",
    "7-isContainedInferred": "",
  },
}
`;
